### PR TITLE
[WIP] Add more tests, fix problems identified by adding more tests

### PR DESCRIPTION
**Pull request by creffett**
Just making the merge request so I can easily see coverall status. I thought I'd be done with tests after graduating college...

### CODE REVIEW

File: test/test.py [hunk 1]
```diff
@@ -41,6 +42,56 @@ def test_handle_nick(self):
         calls = self.send_msg('nick', 'testnick', 'testnick2')
         self.assertEqual(calls, [('testnick', '#test-channel', 0, 'testnick2', 'nick'), ('testnick', '#test-channel2', 0, 'testnick2', 'nick')])
 
+    def test_handle_welcome(self):
+        """Test the bot's ability to handle welcome messages"""
+        self.bot.config['core']['host'] = 'localhost.localhost'  # Override the config value
```
Comment (pefoley2) [inline test/test.py]: If these are more generally useful, they should go in setUpClass()

File: cslbot/commands/coin.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ def cmd(send, msg, _):
     coin = ['heads', 'tails']
     if not msg:
         send('The coin lands on... %s' % choice(coin))
-    elif not msg.isdigit():
+    elif not (msg.isdigit() or (msg[0] == '-' and msg[1:].isdigit())):
```
Comment (pefoley2) [inline cslbot/commands/coin.py]: msg.lstrip('-').isdigit()

File: test/test.py [hunk 3]
```diff
@@ -41,6 +42,56 @@ def test_handle_nick(self):
...
+        self.bot.config['core']['extrachans'] = '#test-channel2'  # Override the config value
+        # Set up buffer capture
+        self.buffer = StringIO()
```
Comment (pefoley2) [inline test/test.py]: use self.assertLogs('handler')

File: test/test.py [hunk 4]
```diff
@@ -41,6 +42,56 @@ def test_handle_nick(self):
...
+        self.logger.removeHandler(self.log_handler)
+        self.log_handler.flush()
+        self.buffer.flush()
+        self.assertLogs(self.log_handler, logging.INFO)
+        self.assertEqual(self.buffer.getvalue(), 'Connected to server localhost.localhost\n')
+        self.assertEqual(calls, [])  # It appears that this shouldn't produce any messages?
+
+    def test_handle_whospcrpl(self):
+        """Test the bot's ability to handle whospcrpl (special who messages)"""
+        self.bot.handler.who_map = {123: '#test-channel'}
```
Comment (pefoley2) [inline test/test.py]: Might want to make this set by a handle_join?

File: test/test.py [hunk 5]
```diff
@@ -64,6 +115,23 @@ def test_morse_encode(self):
         calls = self.send_msg('pubmsg', 'testnick', '#test-channel', ['!morse bob'])
         self.assertEqual(calls, [('testBot', '#test-channel', 0, '-... --- -...', 'privmsg'), ('testnick', '#test-channel', 0, '!morse bob', 'pubmsg')])
 
+    @mock.patch('cslbot.commands.morse.gen_word')
+    def test_morse_noarg(self, mock_gen_word):
+        """Test morse with no arguments"""
+        mock_gen_word.return_value = 'test'
```
Comment (pefoley2) [inline test/test.py]: with mock.patch('gen_word', mock.Mock(return_value='test')):

File: test/test.py [hunk 6]
```diff
@@ -146,6 +197,121 @@ def test_wisdom_search_author_invalid(self):
...
+    def test_definition_invalid_index(self, mock_get):
+        """Test an invalid definition index"""
+        with open(join(dirname(__file__), 'data', 'define_potato.xml')) as test_data_file:
+            mock_get.return_value = mock.Mock(content=test_data_file.read().encode())
+        calls = self.send_msg('pubmsg', 'testnick', '#test-channel', ['!define potato --entry 5'])
+        self.assertEqual(calls,
+                         [('testBot', '#test-channel', 0, 'Invalid index 5 for term potato', 'privmsg'),
+                          ('testnick', '#test-channel', 0, '!define potato --entry 5', 'pubmsg')])
+
+    @mock.patch('cslbot.helpers.misc.choice')
```
Comment (pefoley2) [inline test/test.py]: Split out into ErrnoTest

File: test/test.py [hunk 7]
```diff
@@ -146,6 +197,121 @@ def test_wisdom_search_author_invalid(self):
...
+        self.assertEqual(calls, [('testBot', '#test-channel', 0,
+                                  'EACCES, EADDRINUSE, EADDRNOTAVAIL, EADV, EAFNOSUPPORT, EAGAIN, EALREADY, EBADE, EBADF, EBADFD, EBADMSG, ' +
+                                  'EBADR, EBADRQC, EBADSLT, EBFONT, EBUSY, ECANCELED, ECHILD, ECHRNG, ECOMM, ECONNABORTED, ECONNREFUSED, ECONNRESET, ' +
+                                  'EDEADLK, EDESTADDRREQ, EDOM, EDOTDOT, EDQUOT, EEXIST, EFAULT, EFBIG, EHOSTDOWN, EHOSTUNREACH, EHWPOISON, EIDRM, ' +
+                                  'EILSEQ, EINPROGRESS, EINTR, EINVAL, EIO, EISCONN, EISDIR, EISNAM, EKEYEXPIRED, EKEYREJECTED,', 'privmsg'),
+                                 ('testBot', '#test-channel', 0,
+                                  'EKEYREVOKED, ELIBACC, ELIBBAD, ELIBEXEC, ELIBMAX, ELIBSCN, ELNRNG, ELOOP, EMEDIUMTYPE, EMFILE, EMLINK, EMSGSIZE, ' +
+                                  'EMULTIHOP, ENAMETOOLONG, ENAVAIL, ENETDOWN, ENETRESET, ENETUNREACH, ENFILE, ENOANO, ENOBUFS, ENOCSI,...', 'privmsg'),
+                                 ('testnick', '#test-channel', 0, '!errno list', 'pubmsg')])
+
```
Comment (pefoley2) [inline test/test.py]: SignalTest

File: test/test.py [hunk 8]
```diff
@@ -146,6 +197,121 @@ def test_wisdom_search_author_invalid(self):
...
+    def test_signal_valid(self):
+        """Test signal, basic check only since errno covers most of the backend"""
+        calls = self.send_msg('pubmsg', 'testnick', '#test-channel', ['!signal 9'])
+        self.assertEqual(calls, [('testBot', '#test-channel', 0, '#define SIGKILL 9', 'privmsg'), ('testnick', '#test-channel', 0, '!signal 9', 'pubmsg')])
+
+    @mock.patch('cslbot.commands.coin.choice')
```
Comment (pefoley2) [inline test/test.py]: CoinTest

File: test/test.py [hunk 9]
```diff
@@ -146,6 +197,121 @@ def test_wisdom_search_author_invalid(self):
...
+        self.assertEqual(calls, [('testBot', '#test-channel', 0, 'Negative Flipping requires the (optional) quantum coprocessor.', 'privmsg'),
+                                 ('testnick', '#test-channel', 0, '!coin -1', 'pubmsg')])
+
+    def test_coin_zero(self):
+        """Test coin flipping with arguments"""
+        calls = self.send_msg('pubmsg', 'testnick', '#test-channel', ['!coin 0'])
+        self.assertEqual(calls, [('testBot', '#test-channel', 0, 'The coins land on heads 0 times and on tails 0 times.', 'privmsg'),
+                                 ('testnick', '#test-channel', 0, '!coin 0', 'pubmsg')])
+
+    def test_botsnack_valid_noargs(self):
```
Comment (pefoley2) [inline test/test.py]: BotsnackTest

**Comment by pefoley2** [comment]
Made a bunch of comments, let me know when you want me to do another pass.

File: test/test.py (hunk 4 above)
Comment (creffett) [inline test/test.py]: Not seeing the advantage, since this is pretty specific to whospcrpl tests

File: test/test.py (hunk 5 above)
Comment (creffett) [inline test/test.py]: Not seeing the advantage over the current setup

File: test/bot_test.py [hunk 10]
```diff
@@ -100,6 +103,7 @@ def join_channel(self, nick, channel):
             expected_calls.append((nick, 'private', 0, 'Joined channel %s' % channel, 'privmsg'))
         self.assertEqual(calls, expected_calls)
         self.log_mock.reset_mock()
+        self.raw_mock.reset_mock()
```
Comment (pefoley2) [inline test/bot_test.py]: This really shouldn't go here, the only reason log_mock is reset here is because the send_msg call affects it.

**Pull request merged**